cmd/nerdctl/compose_images.go

```diff
@@ -0,0 +1,190 @@
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"text/tabwriter"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/pkg/progress"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/containerd/nerdctl/pkg/imgutil"
+	"github.com/containerd/nerdctl/pkg/labels"
+	"github.com/containerd/nerdctl/pkg/strutil"
+	"github.com/spf13/cobra"
+	"golang.org/x/sync/errgroup"
+)
+
+func newComposeImagesCommand() *cobra.Command {
+	var composeImagesCommand = &cobra.Command{
+		Use:           "images [flags] [SERVICE...]",
+		Short:         "List images used by created containers in services",
+		RunE:          composeImagesAction,
+		SilenceUsage:  true,
+		SilenceErrors: true,
+	}
+	composeImagesCommand.Flags().BoolP("quiet", "q", false, "Only show numeric image IDs")
+
+	return composeImagesCommand
+}
+
+func composeImagesAction(cmd *cobra.Command, args []string) error {
+	quiet, err := cmd.Flags().GetBool("quiet")
+	if err != nil {
+		return err
+	}
+
+	client, ctx, cancel, err := newClient(cmd)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	c, err := getComposer(cmd, client)
+	if err != nil {
+		return err
+	}
+
+	serviceNames, err := c.ServiceNames(args...)
+	if err != nil {
+		return err
+	}
+
+	containers, err := c.Containers(ctx, serviceNames...)
+	if err != nil {
+		return err
+	}
+
+	if quiet {
+		return printComposeImageIDs(ctx, containers)
+	}
+
+	snapshotter, err := cmd.Flags().GetString("snapshotter")
+	if err != nil {
+		return err
+	}
+	sn := client.SnapshotService(snapshotter)
+
+	return printComposeImages(ctx, cmd, containers, sn)
+}
+
+func printComposeImageIDs(ctx context.Context, containers []containerd.Container) error {
+	ids := []string{}
+	for _, c := range containers {
+		image, err := c.Image(ctx)
+		if err != nil {
+			return fmt.Errorf("Failed to get image for container %s: %s", c.ID(), err.Error())
+		}
+		metaImage := image.Metadata()
+		id := metaImage.Target.Digest.String()
+		if !strutil.InStringSlice(ids, id) {
+			ids = append(ids, id)
+		}
+	}
+
+	for _, id := range ids {
+		// always truncate image ids.
+		fmt.Println(strings.Split(id, ":")[1][:12])
+	}
+	return nil
+}
+
+func printComposeImages(ctx context.Context, cmd *cobra.Command, containers []containerd.Container, sn snapshots.Snapshotter) error {
+	type composeImagePrintable struct {
+		ContainerName string
+		Repository    string
+		Tag           string
+		ImageID       string
+		Size          string
+	}
+
+	var (
+		imagePrintables = []composeImagePrintable{}
+		mu              sync.Mutex
+	)
+
+	eg, ctx := errgroup.WithContext(ctx)
+	for _, c := range containers {
+		c := c
+		eg.Go(func() error {
```

Comment:
not sure if the `errorgroup goroutines` here are over-engineering/a plain for-loop is sufficient. Thanks for any suggestion :)

Reply:
errgroup goroutines look good